core/src/main/scala/org/apache/spark/sql/crossdata/XDContext.scala

```diff
@@ -126,45 +127,49 @@ class XDContext private (@transient val sc: SparkContext,
 
       override lazy val batches: Seq[Batch] = Seq(
         Batch("Substitution", fixedPoint,
-          CTESubstitution ::
-            WindowsSubstitution ::
-            Nil : _*),
+          CTESubstitution,
+          WindowsSubstitution),
         Batch("Preparation", fixedPoint, preparationRules : _*),
         Batch("Resolution", fixedPoint,
           ResolveRelations ::
             ResolveReferences ::
             ResolveGroupingAnalytics ::
+            ResolvePivot ::
+            ResolveUpCast ::
             ResolveSortReferences ::
             ResolveGenerate ::
             ResolveFunctions ::
             ResolveAliases ::
             ExtractWindowExpressions ::
             GlobalAggregates ::
             ResolveAggregateFunctions ::
+            DistinctAggregationRewriter(conf) ::
             HiveTypeCoercion.typeCoercionRules ++
               extendedResolutionRules : _*),
         Batch("Nondeterministic", Once,
-          PullOutNondeterministic),
+          PullOutNondeterministic,
+          ComputeCurrentTime),
+        Batch("UDF", Once,
+          HandleNullInputsForUDF),
         Batch("Cleanup", fixedPoint,
           CleanupAliases)
       )
     }
 
   @transient
-  class XDPlanner extends SparkPlanner with XDStrategies {
+  class XDPlanner extends sparkexecution.SparkPlanner(this) with XDStrategies {
```

Comment:
SparkPlanner is an inner class within SQLContext (although deprecated), so it should be changed in Spark 2.0 when the class is removed